crates/core/src/factories/core_kitsune.rs

```diff
@@ -170,6 +170,101 @@ impl Kitsune for CoreKitsune {
         })
     }
 
+    fn remove_space(&self, space: SpaceId) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
+            let fut = self.map.lock().unwrap().get(&space).cloned();
+            if let Some(fut) = fut {
+                if let Ok(s) = fut.await {
+                    // Just a point in time check, try to help the user out if they accidentally
+                    // try to remove a space with active local agents.
+                    // Those agents need to leave the space before it should be removed.
+                    if !s.local_agent_store().get_all().await?.is_empty() {
+                        return Err(K2Error::other(
+                            "Cannot remove space with local agents",
+                        ));
+                    }
+
+                    // Checks passed, remove our reference to the space.
+                    self.map.lock().unwrap().remove(&space);
+
+                    // Unregister the space and its module handlers from the transport.
+                    self.transport().await?.unregister_space(space).await;
+
+                    // Get all the peer URLs of connected peers.
+                    let connected_peer_urls = Arc::new(RwLock::new(
+                        self.transport()
+                            .await?
+                            .get_connected_peers()
+                            .await?
+                            .into_iter()
+                            .collect::<HashSet<_>>(),
+                    ));
+
+                    // Remove any peer URLs that are connected for another space.
+                    let spaces = self
+                        .map
+                        .lock()
+                        .unwrap()
+                        .iter()
+                        .map(|s| s.1.clone())
+                        .collect::<Vec<_>>();
+                    futures::future::join_all(spaces.into_iter().map(
+                        |fut| -> BoxFut<'_, K2Result<()>> {
+                            let connected_peer_urls =
+                                connected_peer_urls.clone();
+                            Box::pin(async move {
+                                if let Ok(s) = fut.await {
+                                    let check_our_peers = connected_peer_urls
+                                        .read()
+                                        .unwrap()
+                                        .clone();
+                                    let space_agent_infos =
+                                        s.peer_store().get_all().await?;
+                                    for info in space_agent_infos {
+                                        if let Some(url) = &info.url {
+                                            if check_our_peers.contains(url) {
```

Comment:
Is it better to filter space_agent_infos and then remove all remaining urls under one write lock?

Reply:
I don't know how I'd do that. The lock is local to this function and only created so that Rust is happy that we're not modifying the `HashSet` from multiple threads while these futures execute in parallel